File: web/src/types/session.ts
export type AgentType = 'claude' | 'codex';
export type SessionType = 'agent' | 'shell';
export type SessionStatus = 'pending' | 'starting' | 'running' | 'idle' | 'stopping' | 'stopped' | 'failed';

export interface Session {
  id: string;
  name: string;
  host_id: string;
  template_id: string;
  type: SessionType;
  agent_type?: AgentType;
  status: SessionStatus;
  prompt?: string;
  branch_name?: string;
  worktree_path?: string;
  working_directory?: string;  // Custom working directory (alternative to worktree)
  tmux_session: string;
  agent_pid: number;
  model_override?: string;
  env_vars: Record<string, string>;
  mcp_servers: unknown[];
  plugins: string[];
  allowed_tools: string[];
  extra_flags: string[];
  last_output: string;
  started_at: string | null;
  stopped_at: string | null;
  created_at: string;
  updated_at: string;
}

export type DirectorySourceType = 'existing' | 'new_folder' | 'clone_repo' | 'new_repo';

export interface DirectorySource {
  type: DirectorySourceType;
  existing_path?: string;       // For type='existing'
  new_folder_name?: string;      // For type='new_folder'
  repo_url?: string;             // For type='clone_repo'
  repo_name?: string;            // For type='new_repo'
  repo_description?: string;     // For type='new_repo'
  repo_private?: boolean;        // For type='new_repo'
  clone_folder_name?: string;    // Optional custom folder name for cloned repos
}

export interface CreateSessionRequest {
  name?: string;
  host_id: string;
  type?: SessionType;
  agent_type?: AgentType;
  prompt?: string;
  branch_name?: string;
  template_id?: string;
  model_override?: string;
  env_vars?: Record<string, string>;
  plugins?: string[];
  allowed_tools?: string[];
  extra_flags?: string[];
  working_directory?: string;    // Custom working directory path
  directory_source?: DirectorySource;  // How to set up the working directory
}

File: pkg/models/session.go
package models

import "time"

type AgentType string

const (
	AgentTypeClaude AgentType = "claude"
	AgentTypeCodex  AgentType = "codex"
)

type SessionType string

const (
	SessionTypeAgent SessionType = "agent" // Agent-based session (claude/codex in a repo)
	SessionTypeShell SessionType = "shell" // Interactive shell session
)

type SessionStatus string

const (
	SessionStatusPending  SessionStatus = "pending"
	SessionStatusStarting SessionStatus = "starting"
	SessionStatusRunning  SessionStatus = "running"
	SessionStatusIdle     SessionStatus = "idle"
	SessionStatusStopping SessionStatus = "stopping"
	SessionStatusStopped  SessionStatus = "stopped"
	SessionStatusFailed   SessionStatus = "failed"
)

type DirectorySourceType string

const (
	DirectorySourceExisting  DirectorySourceType = "existing"
	DirectorySourceNewFolder DirectorySourceType = "new_folder"
	DirectorySourceCloneRepo DirectorySourceType = "clone_repo"
	DirectorySourceNewRepo   DirectorySourceType = "new_repo"
)

// DirectorySource describes how to set up the working directory for a session
type DirectorySource struct {
	Type             DirectorySourceType `json:"type"`
	ExistingPath     string              `json:"existing_path,omitempty"`
	NewFolderName    string              `json:"new_folder_name,omitempty"`
	RepoURL          string              `json:"repo_url,omitempty"`
	RepoName         string              `json:"repo_name,omitempty"`
	RepoDescription  string              `json:"repo_description,omitempty"`
	RepoPrivate      bool                `json:"repo_private,omitempty"`
	CloneFolderName  string              `json:"clone_folder_name,omitempty"`
}

type Session struct {
	ID               string            `json:"id" db:"id"`
	Name             string            `json:"name" db:"name"`
	HostID           string            `json:"host_id" db:"host_id"`
	TemplateID       string            `json:"template_id,omitempty" db:"template_id"`
	Type             SessionType       `json:"type" db:"type"`
	AgentType        AgentType         `json:"agent_type,omitempty" db:"agent_type"`
	Status           SessionStatus     `json:"status" db:"status"`
	Prompt           string            `json:"prompt,omitempty" db:"prompt"`
	BranchName       string            `json:"branch_name,omitempty" db:"branch_name"`
	WorktreePath     string            `json:"worktree_path,omitempty" db:"worktree_path"`
	WorkingDirectory string            `json:"working_directory,omitempty" db:"working_directory"` // Custom working directory (alternative to worktree)
	TmuxSessionName  string            `json:"tmux_session" db:"tmux_session"`
	AgentPID         int               `json:"agent_pid" db:"agent_pid"`
	ModelOverride    string            `json:"model_override,omitempty" db:"model_override"`
	EnvVars          map[string]string `json:"env_vars"`
	MCPServers       []MCPServerConfig `json:"mcp_servers"`
	Plugins          []string          `json:"plugins"`
	AllowedTools     []string          `json:"allowed_tools"`
	ExtraFlags       []string          `json:"extra_flags"`
	LastOutput       string            `json:"last_output,omitempty" db:"last_output"`
	StartedAt        *time.Time        `json:"started_at,omitempty" db:"started_at"`
	StoppedAt        *time.Time        `json:"stopped_at,omitempty" db:"stopped_at"`
	CreatedAt        time.Time         `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time         `json:"updated_at" db:"updated_at"`
}
